specs/governance/alligator-interop.md

```diff
@@ -0,0 +1,136 @@
+
+## Interface
+
+### Core Functions
+
+```solidity
+function afterTokenTransfer(
+    address src,
+    address dst,
+    uint256 amount
+) external
+```
+
+In addition to performing it's standard functionality this method will now use the `L2ToL2CrossDomainMessenger` to send a message to a new method on the `Alligator` contract, `afterTokenTransferInterop`. It will ensure prior to doing so that the `chainId` is not OP Mainnet i.e. 10(0xa). Moreover, it will still handle a check that both address have been migrated using
+the new storage field and update the voting power of the delegate.
+
+```solidity
+function afterTokenTransferInterop(
+    address src,
+    address dst,
+    uint256 amount
+) external
+```
+
+The new function that will be added to the `Alligator` contract will process the interop messages being sent by the `afterTokenTransfer` method. The only address that can call
+this method is the `L2ToL2CrossChainMessenger`. Furthermore, it will check that the sender of this message is in fact another OP token with the same address, which is assumed since the
+goverance token is a predeploy. Given that messages are received from other tokens that exist on the superchain, an additional check needs to be performed that ensures we are on OP
+mainnet by checking the `chainId` before updating the checkpoints for the delegate voting power.
+
+## Diagram
+
+```mermaid
+sequenceDiagram
+  participant delegate
+  participant AlligatorSuperChain as AlligatorSuperChain (Chain A)
+  participant AlligatorOPMainnet as AlligatorOPMainnet (Chain B)
+  participant Inbox as CrossL2Inbox
+  participant Messenger_A as L2ToL2CrossDomainMessenger (Chain A)
+  participant Messenger_B as L2ToL2CrossDomainMessenger (Chain B)
+
+  delegate->>AlligatorSuperChain: afterTokenTransfer(src, dst, amt)
+  AlligatorSuperChain->>Messenger_A: sendMessage(nativeChainId, message)
+  Messenger_A->>Inbox: executeMessage()
+  Inbox->>Messenger_B: relayMessage()
+  Messenger_B->>AlligatorOPMainnet: afterTokenTransferInterop(src, dst, amount, chainId)
+```
+
+## Implementation
+
+```solidity
+function afterTokenTransfer(
+    address src,
+    address dst,
+    uint256 amount
+) external {
+    if (src != dst && amount > 0) {
+       if (src != address(0)) {
+            (uint256 oldWeight, uint256 newWeight) = _writeCheckpoint(_checkpoints[src], _subtract, amount);
+            emit DelegateVotesChanged(src, oldWeight, newWeight);
+        }
+
+        if (dst != address(0)) {
+            (uint256 oldWeight, uint256 newWeight) = _writeCheckpoint(_checkpoints[dst], _add, amount);
+            emit DelegateVotesChanged(dst, oldWeight, newWeight);
+        }
+
+        uint256 nativeChainId;
+        assembly {
+             nativeChainId := chainid()
+        }
+
+        if (nativeChainId != uint256(10)) {
```

Comment:
This is not going to be scalable as every token transfer on a remote chain will require a relayer to send a transaction on op mainnet. This means gov will need to front the costs here because nobody else has the incentive to do so. Wondering if there is a way that we can:
- send batched diffs
- not need to be aware of the chainid

Going to brainstorm a bit on this